### PR TITLE
feat(aws): Adds support for AWS MSK

### PR DESCRIPTION
**Pull request by robh007**
**Objective**:

Adds support for ```msk_cluster```

** Closes ** #205 

**Example output:**

```
NAME                                      MONTHLY QTY  UNIT       PRICE   HOURLY COST  MONTHLY COST  

  aws_msk_cluster.test                                                                                 
  ├─ MSK (broker instance, kafka.t3.small)        2,920  hours      0.0456       0.1824      133.1520  
  └─ MSK storage                                  4,000  GB-months  0.1000       0.5479      400.0000  
  Total                                                                          0.7303      533.1520  
                                                                                                       
  OVERALL TOTAL (USD)                                                            0.7303      533.1520  


```

**Pricing details:**

The Amazon MSK service provides 2 pricing components described in https://aws.amazon.com/msk/pricing/. This adds support for the Amazon MSK Clusters.

### CODE REVIEW

File: internal/providers/terraform/aws/msk_cluster.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+	brokerNodes := decimal.NewFromInt(d.Get("number_of_broker_nodes").Int())
+	brokerInstance := d.Get("broker_node_group_info.0.instance_type").String()
+	ebsVolumeSize := decimal.NewFromInt(d.Get("broker_node_group_info.0.ebs_volume_size").Int()).Mul(brokerNodes)
+
+	return &schema.Resource{
+		Name: d.Address,
+		CostComponents: []*schema.CostComponent{
+			{
+				Name:           fmt.Sprintf("MSK (broker instance, %s)", brokerInstance),
```
Comment (alikhajeh1) [inline internal/providers/terraform/aws/msk_cluster.go:28]:
```suggestion
				Name:           fmt.Sprintf("Instance (%s)", instanceType),
```

File: internal/providers/terraform/aws/msk_cluster.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		Name:  "aws_msk_cluster",
+		RFunc: NewMskCluster,
+	}
+}
+
+func NewMskCluster(d *schema.ResourceData, u *schema.ResourceData) *schema.Resource {
+	region := d.Get("region").String()
+
+	brokerNodes := decimal.NewFromInt(d.Get("number_of_broker_nodes").Int())
+	brokerInstance := d.Get("broker_node_group_info.0.instance_type").String()
```
Comment (alikhajeh1) [inline internal/providers/terraform/aws/msk_cluster.go:21]:
```suggestion
	instanceType := d.Get("broker_node_group_info.0.instance_type").String()
```

File: internal/providers/terraform/aws/msk_cluster.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+				Unit:           "hours",
+				UnitMultiplier: 1,
+				HourlyQuantity: &brokerNodes,
+				ProductFilter: &schema.ProductFilter{
+					VendorName:    strPtr("aws"),
+					Region:        strPtr(region),
+					Service:       strPtr("AmazonMSK"),
+					ProductFamily: strPtr("Managed Streaming for Apache Kafka (MSK)"),
+					AttributeFilters: []*schema.AttributeFilter{
+						{Key: "computeFamily", ValueRegex: strPtr(fmt.Sprintf("/%s/", strings.SplitN(brokerInstance, ".", 2)[1]))},
```
Comment (alikhajeh1) [inline internal/providers/terraform/aws/msk_cluster.go:38]:
```suggestion
						{Key: "locationType", Value: strPtr("AWS Region")},
						{Key: "usagetype", ValueRegex: strPtr(fmt.Sprintf("/%s/", instanceType))},
```

File: internal/providers/terraform/aws/msk_cluster.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+					},
+				},
+			},
+			{
+				Name:            "MSK storage",
```
Comment (alikhajeh1) [inline internal/providers/terraform/aws/msk_cluster.go:43]:
```suggestion
				Name:            "Storage",
```

**Review by alikhajeh1 — COMMENTED** [review]
LGTM, a few minor suggestions.

I wonder if we should add `{Key: "locationType", Value: strPtr("AWS Region")},` to these new resources so if/when AWS adds Outpost prices, we don't need to change things on our side. Any thoughts?

File: internal/providers/terraform/aws/msk_cluster.go (hunk 1 above)
Comment (robh007) [reply internal/providers/terraform/aws/msk_cluster.go:28]: It makes sense to add the locationType.

**Comment by robh007** [comment]
@alikhajeh1 Updated as per your comments.

File: internal/providers/terraform/aws/msk_cluster.go (hunk 2 above)
Comment (robh007) [reply internal/providers/terraform/aws/msk_cluster.go]: The price component uses "Kafka" so I've added a string replace to uppercase the first letter. I can't use Title as the instance contains "." and produces "Kafka.T3.Small" which fails to retrieve any prices.

File: internal/providers/terraform/aws/msk_cluster.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+		Name:  "aws_msk_cluster",
+		RFunc: NewMskCluster,
+	}
+}
+
+func NewMskCluster(d *schema.ResourceData, u *schema.ResourceData) *schema.Resource {
+	region := d.Get("region").String()
+
+	brokerNodes := decimal.NewFromInt(d.Get("number_of_broker_nodes").Int())
+	instanceType := strings.ReplaceAll(d.Get("broker_node_group_info.0.instance_type").String(), "kafka", "Kafka")
```
Comment (alikhajeh1) [inline internal/providers/terraform/aws/msk_cluster.go:21]:
```suggestion
	instanceType := d.Get("broker_node_group_info.0.instance_type").String()
```

File: internal/providers/terraform/aws/msk_cluster.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+				Unit:           "hours",
+				UnitMultiplier: 1,
+				HourlyQuantity: &brokerNodes,
+				ProductFilter: &schema.ProductFilter{
+					VendorName:    strPtr("aws"),
+					Region:        strPtr(region),
+					Service:       strPtr("AmazonMSK"),
+					ProductFamily: strPtr("Managed Streaming for Apache Kafka (MSK)"),
+					AttributeFilters: []*schema.AttributeFilter{
+						{Key: "usagetype", ValueRegex: strPtr(fmt.Sprintf("/%s/", instanceType))},
```
Comment (alikhajeh1) [inline internal/providers/terraform/aws/msk_cluster.go:38]:
```suggestion
						{Key: "usagetype", ValueRegex: strPtr(fmt.Sprintf("/%s/i", instanceType))},
```

**Review by alikhajeh1 — APPROVED** [review]
ah didn't notice the case sensitive issue, I tweaked the code to do a case-insensitive regex, if it's ok with you I'll commit and merge those?

File: internal/providers/terraform/aws/msk_cluster.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
+package aws
+
+import (
+	"fmt"
+	"github.com/infracost/infracost/internal/schema"
+	"github.com/shopspring/decimal"
+	"strings"
```
Comment (alikhajeh1) [inline internal/providers/terraform/aws/msk_cluster.go:7]:
```suggestion
```

File: internal/providers/terraform/aws/msk_cluster_test.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+             security_groups = []
+           }
+       }
+`
+	resourceChecks := []testutil.ResourceCheck{
+		{
+			Name: "aws_msk_cluster.cluster-2-nodes",
+			CostComponentChecks: []testutil.CostComponentCheck{
+				{
+					Name:             "Instance (Kafka.t3.small)",
```
Comment (alikhajeh1) [inline internal/providers/terraform/aws/msk_cluster_test.go:45]:
```suggestion
					Name:             "Instance (kafka.t3.small)",
```

File: internal/providers/terraform/aws/msk_cluster_test.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+					Name:             "Storage",
+					PriceHash:        "c051f24392b78290de9613ef486ba264-ee3dd7e4624338037ca6fea0933a662f",
+					MonthlyCostCheck: testutil.MonthlyPriceMultiplierCheck(decimal.NewFromInt(1000)),
+				},
+			},
+		}, {
+			Name: "aws_msk_cluster.cluster-4-nodes",
+			CostComponentChecks: []testutil.CostComponentCheck{
+				{
+					Name:             "Instance (Kafka.m5.24xlarge)",
```
Comment (alikhajeh1) [inline internal/providers/terraform/aws/msk_cluster_test.go:59]:
```suggestion
					Name:             "Instance (kafka.m5.24xlarge)",
```

**Pull request merged**